etl/extractor/mainnet_impl.go

```diff
@@ -71,32 +175,62 @@ func (m *MainNetExtractor) GetJetDrops(ctx context.Context) <-chan *types.Platfo
 				}
 				resp, err := stream.Recv()
 				if err == io.EOF {
-					// log.Debug("EOF received, quit")
-					println("EOF received, quit")
+					log.Debug("EOF received, quit")
+					closeStream(ctx, stream)
 					break
 				}
 				if err != nil {
-					// log.Debug("received error value from records gRPC stream %v", m.request)
-					println("received error value from records gRPC stream %v", m.request)
+					log.Debug("received error value from records gRPC stream %v", m.request)
 					errorChan <- errors.Wrapf(err, "received error value from records gRPC stream %v", m.request)
 				}
 
+				if resp.ShouldIterateFrom != nil {
+					m.request.PulseNumber = *resp.ShouldIterateFrom
+					receivedPulseNumber = m.request.PulseNumber.AsUint32()
+					break
+				}
+
 				// save the last pulse for future requests
 				m.request.RecordNumber = resp.RecordNumber
 				m.request.PulseNumber = resp.Record.ID.Pulse()
+				receivedPulseNumber = m.request.PulseNumber.AsUint32()
+
+				// collect all records by PulseNumber
+				if receivedPulseNumber == lastPulseNumber {
+					jetDrops.Records = append(jetDrops.Records, resp)
+					continue
+				}
+
+				lastPulseNumber = receivedPulseNumber
 
-				jetDrops := new(types.PlatformJetDrops)
-				jetDrops.Records = append(jetDrops.Records, resp)
 				m.mainJetDropsChan <- jetDrops
+				// zeroing variable which collecting jetDrops
+				jetDrops = new(types.PlatformJetDrops)
+				// don't forget to save the last data
+				jetDrops.Records = append(jetDrops.Records, resp)
 			}
 		}
 	}()
+}
 
-	return m.mainJetDropsChan
+func closeStream(ctx context.Context, stream exporter.RecordExporter_ExportClient) {
+	streamError := stream.CloseSend()
+	if streamError != nil {
+		belogger.FromContext(ctx).Warn("Error closing stream: ", streamError)
+	}
 }
 
-func (m *MainNetExtractor) Stop() {
-	m.stopSignal <- true
+var startStopMutes = &sync.Mutex{}
```

Comment:
Why it variable and not `MainNetExtractor` field? It seems to protect specific MainNetExtractor instance